### PR TITLE
Refactor: separate ctxrunner from database

### PR DESCRIPTION
**Pull request by takluyver**
This continues the work I started in #192. My goal this time was to separate the low-level ctxrunner machinery, which parses and runs a `context.py` file, from the database.

Using user-editable variables in the context file has never been turned on, and this PR still doesn't, but it changes the way they will work if/when we do. They would be labelled with an `input#` prefix (e.g. `input#sample_type`) instead of `var#` like variables defined in the context file. A higher level piece would be responsible for finding their values and passing them in when running a context file - the precise mechanism for passing them down still to be decided.

I also moved some of the context file checks into a separate `ContextFile.check()` method. The idea is that we can still read a context file and tell you about its contents even if e.g. a raw variable depends on a proc variable. For now, the context file is checked whenever we execute it, so if there are any issues all processing is stopped, but I'd like to relax that in the future, e.g. to allow processing variables without issues.

### CODE REVIEW

**Review by JamesWrigley — CHANGES_REQUESTED** [review]
I appreciate the commit organization :heart:

File: damnit/ctxsupport/ctxrunner.py [hunk 1]
```diff
@@ -167,52 +154,57 @@ def filter(self, run_data=RunData.ALL, cluster=True, name_matches=()):
...
+                    log.warning(f"Skipping {name} because of missing dependencies: {', '.join(missing_deps)}")
+                    continue
+                elif missing_input:
+                    log.warning(f"Skipping {name} because of missing input variables: {', '.join(missing_input)}")
+                    continue
+
                 func = functools.partial(var.func, **kwargs)
 
-                data = func(inputs)
+                data = func(run_data)  # TODO - fix methods
```
[the remark this answers is not in the document]
Comment (takluyver) [reply damnit/ctxsupport/ctxrunner.py:229]: It's now `# DONE` :slightly_smiling_face: . When making this change, I left myself a note to make the corresponding change in `damnit_ctx`, then did that and forgot to remove the note.

```suggestion
                data = func(run_data)
```

File: damnit/ctxsupport/ctxrunner.py [hunk 2]
```diff
@@ -370,27 +392,27 @@ def save_hdf5(self, hdf5_path, reduced_only=False):
 
         xarray_dsets = []
         obj_type_hints = {}
-        dsets = [(f'.reduced/{name}', v) for name, v in self.reduced.items()
-                 if name in implicit_vars or ctx_vars[name].store_result]
+        dsets = [(f'.reduced/{name}', v) for name, v in self.reduced.items()]
+                 #if name in implicit_vars or ctx_vars[name].store_result]
```
[the remark this answers is not in the document]
Comment (takluyver) [reply damnit/ctxsupport/ctxrunner.py:396]: Yup, this was finding only the variables that were computed. I commented it out to test and forgot to clean it up, thanks!

```suggestion
```

File: damnit/ctxsupport/ctxrunner.py [hunk 3]
```diff
@@ -397,27 +392,27 @@ def save_hdf5(self, hdf5_path, reduced_only=False):
...
         if not reduced_only:
             for name, obj in self.data.items():
-                if name in implicit_vars or ctx_vars[name].store_result:
-                    if isinstance(obj, (xr.DataArray, xr.Dataset)):
-                        xarray_dsets.append((name, obj))
-                        obj_type_hints[name] = (
-                            DataType.DataArray if isinstance(obj, xr.DataArray)
-                            else DataType.Dataset
-                        )
+                # if name in implicit_vars or ctx_vars[name].store_result:
```
Comment (takluyver) [inline damnit/ctxsupport/ctxrunner.py:399]:
```suggestion
```

**Comment by JamesWrigley** [comment]
Ah joy :grimacing: 
image

That happened once to me this morning too, not sure. Let's fix it in another PR unless it really messes with CI.

**Comment by JamesWrigley** [comment]
(oh, I missed that the `test_context_file` test is actually failing, maybe it's related to that)

**Review by JamesWrigley — APPROVED** [review]
LGTM!

**Comment by takluyver** [comment]
Thanks James!

**Pull request merged**